Add BLS libraries, wrappers, helpers and benchmarks

codablock opened the pull request:
**UPDATE: Update description to only refer to the Chia BLS library** 

This adds the BLS librariy https://github.com/Chia-Network/bls-signatures to depends.

It also adds the following things:
1. Wrappers around the Chia lib so that we can stay independent from thes library. These wrappers also add Dash compatible serialization.
2. A simple BLS+Diffie-Hellmann+AES-256CBC integrated encryption scheme implementation
3. A highly paralellized worker/helper for all kinds of BLS related operations. These leverage aggregation and the "correlation" property of BLS as much as possible to speed up things.
4. Add the ctpl header-only library which is used by the parallel BLS helper/worker.
5. Benchmarks for BLS and the BLS worker
6. ECDSA benchmarks for comparision

I decided to separate them from the LLMQ work I've done. This allows other people to experiment with the BLS primitives beforehand.

codablock commented:
FYI, the "Add https://github.com/herumi/xxx libs for BLS support" currently contains a patch that adds a few function to the libs. This patch will vanish when/if https://github.com/herumi/bls/pull/13 gets merged.

krishnawhite commented:
Maybe I'm being paranoid, but has anyone reviewed the security of these libraries? They seem to be written almost entirely by one person from a Japanese university, I don't get the feeling that they've been widely reviewed.

codablock commented:
The reason we took these libraries are that DFINITY uses them for their DKG. Another reason is that it already supports JavaScript (WebAssembly based) and many other languages.

But I must agree with you, review is a big problem with these libraries. They seem to be not very well reviewed and are extremely hard to understand. They also contain a lot of stuff that we don't need, making it even harder to review them.

I'm already looking into alternative libraries tbh, e.g. https://github.com/Chia-Network/bls-signatures. What I've seen so far seems to be very clean and easy to understand code. Problem is, that this library does not implement everything that we need (e.g. Shamir's Secret Sharing). It's also unclear if a JavaScript version will be possible and it is limited to the BLS12-381 curve in reversed mode. It's not 100% clear yet which curves we'll use in the end (it might even be that we need multiple curves for different use-cases).

codablock commented:
Force pushed a new version. The BLS wrappers now support the herumi libraries and the Chia BLS library. Herumi can be configured with 2 different curves for now (you have to manually change a few `#if`). Chia only support BLS12-381 in reversed (g1 and g2 swapped) mode.

I'm in contact with the Chia developer to figure out if we can change the library to better fit our needs. We especially need to be able to perform aggregation/verification without the overhead of the secure AggregationInfo. We might also need to swap g1 and g2 back as we actually have downsides because of this (we have more signatures than public keys on-chain). If BLS12-381 turns out to be too slow (and it's really slow...), we might also need to add support for BN curves to Chia, which the developer is open for contributions.

For now, all the changes I needed to do to Chia are in https://github.com/codablock/bls-signatures/tree/dash and I'm trying to PR as much as possible into upstream.

@UdjinM6 If you compile this on mac, it will automatically use the Chia library as it was too much effort to cross-compile the herumi libraries.

codablock commented:
FYI, if you run benchmarks with Chia, it will be extremely slow. This is due to the overhead of the AggregationInfo bookkeeping.

codablock commented:
Updated the herumi libs to latest upstream, which removed most of dash.patch :)

UdjinM6 commented:
As mentioned earlier in Slack, I had to add this patch https://github.com/UdjinM6/dash/commit/c2ef94f6fa06b86d36db4f69d3379785cc9aedb4 (and  #2325) for chia_bls to be successfully compiled/staged/linked on macos.

codablock commented:
Updated Chia libs to latest version, including some refactoring that is not merged yet in the Chia lib. With this, Chia can compete in the benchmarks with the Herumi libs.

Also, build should be fixed on all platforms now...waiting for Travis to confirm this.

UdjinM6 commented on depends/packages/chia_bls.mk:
```diff
@@ -0,0 +1,49 @@
+  export CFLAGS="$($(package)_cflags)" && \
+  export CXXFLAGS="$($(package)_cxxflags)" && \
+  export LDFLAGS="$($(package)_ldflags)" && \
+  mkdir build && cd build && \
```
This fails for me on linux with `"build" already exists` (not sure why, wasn't the case on mac for some reason), should be `mkdir build ; cd build && \` probably.

UdjinM6 replied:
n/m, there was an issue with my local setup

codablock commented:
Thanks for the code-style fixes. I squashed https://github.com/UdjinM6/dash/commit/f8d920e771ee9b743c7c75e37b7edc7e13238e3c directly into the original commits.

Regarding src/crypto, I added it there because I did not consider that crypto only contains low-level stuff. Would you be ok with adding a sub-folder in src just for bls stuff?

UdjinM6 commented on src/bench/bls.cpp:
```diff
@@ -0,0 +1,361 @@
+    msgHashes.resize(count);
+
+    invalid.resize(count);
+    for (size_t i = 0; i < invalidCount; i++) {
```
nit: post-increment here and in a bunch of other loops (not a big deal for simple types but should use pre-increment in general)

codablock replied:
I usually only do pre-increment on iterators and trust in the compiler to properly optimize it for other types

UdjinM6 replied:
Agree, you can ignore this one :)
